File: src/python/periodic_KL.py
import pylab
from numpy import *
from numpy.linalg import *
from scipy import stats

N = 100
th = linspace(0,2*pi,N)

x = cos(th)
y = sin(th)
d = sqrt((x - x[0])**2 + (y - y[0])**2)

sig = 1.0
lam = 0.25
f = sig**2*exp(-d**2/(2*lam**2))

F = fft.fft(f).real/N

Z = random.randn(N)
p = zeros((1,N))
for i in arange(N/2+1):
    p += Z[i]*F[i]*cos(i*th)
for i in arange(1,N/2):
    p += Z[N/2+i]*F[i]*sin(i*th)

pylab.plot(th,p.T)
pylab.plot(th+2*pi,p.T)
pylab.plot(th[-1],p.T[-1],'ro')
pylab.show()

File: src/python/periodic_KL.m
% simulate periodic GP on a circle

close all
clear all

N = 100;
th = linspace(0,2*pi,N);

% th = th(1:end-1); % needed?

x = cos(th);
y = sin(th);
d = sqrt((x - x(1)).^2 + (y - y(1)).^2);

sigma = 1.0;
lambda = 0.25;
f = sigma^2*exp(-d.^2/(2*lambda^2));

% these are the eigenvalues
F = real(fft(f))/N;

% form the K-L expansion
Z = randn(N);
% N = 100: modes 1:51 are cosine, 2:50 are also sine
p = zeros(1,N);
for i = 1:N/2+1
    p = p + Z(i)*F(i)*cos(i*th);
end
for i = 2:N/2
    p = p + Z(i)*F(i)*sin(i*th);
end

plot(th,p)
hold on
plot(th+2*pi,p)
plot(th(end),p(end),'ro')

